### PR TITLE
Regex construction optimization

### PR DESCRIPTION
**Pull request by vhavlena**
This PR introduces:

- intermediate reduction during the regex creation

### CODE REVIEW

**Comment by vhavlena** [comment]
So-far the bottleneck seems to be the concatenation in mata.

**Comment by Adda0** [comment]
What about this PR? Is there a reason we did not merge this PR yet?

**Comment by vhavlena** [comment]
I want to add here the postponed epsilon removal for concatenation

**Comment by vhavlena** [comment]
@jurajsic Can you run some automata-intense benchmarks please? I am curious if it indeed has some impact.

**Comment by vhavlena** [comment]
I would like to move on with this PR. The new concatenation seems to work fine. But the current bottleneck seems to be the complementation as Juraj said. So, I want to look more closely on determinization+complementation in mata.

**Comment by jurajsic** [comment]
Did you rebase it correctly? Because the `add` -> `insert` from new mata should already be in devel I think.

**Comment by vhavlena** [comment]
I have fixed it manually in the last commit

**Comment by vhavlena** [comment]
However, the problem remains the same. I think the lines 523--525 should be removed.

**Comment by vhavlena** [comment]
@jurajsic Could you plese run this version on `automatark` benchmark? I tried to fix some TOs.

**Comment by vhavlena** [comment]
For `stringfuzz` there should be some improvement as well.

**Comment by jurajsic** [comment]
We have 92 TOs on `automatark` benchmark now, do you want to compare with noodler on `devel`?

**Comment by vhavlena** [comment]
Yes please

**Comment by jurajsic** [comment]
The version on current devel has 87 timeouts on `automatark`... Here is the scatter plot, `f46a438` is devel, `d251a10` is this PR:
image

**Comment by vhavlena** [comment]
Thanks. The comparison is not totally fair. Devel contains bugs in regex construction that were fixed in this PR. I hoped that these simple optimizations will have a better impact. In that case I am going to implement handling of simple not in constraints (probably all from automatark) via universality checking (the last complement is difficult in many cases)

**Comment by jurajsic** [comment]
Yeah, I was not sure if the fixed bugs from this PR has somehow ended in devel or not yet. For `stringfuzz` there is some improvement:
image
We still have 609 TOs (compared to 998 in devel) and lots of unknowns (1569). There is also some discrepancy with z3str3RE.

**Comment by vhavlena** [comment]
Do we comply with cvc5 on stringfuzz? Or there is a problem on z3strRE side? Can you send me some of the unknowns?

**Comment by jurajsic** [comment]
Yeah, it is a problem with z3str3RE. Also the unknowns are the same as before, so I think it is some unsupported operation.

**Comment by vhavlena** [comment]
Can I merge it? We can get back to automata optimizations later. There are still 64 TOs in the automatark benchmark. I want to switch to something else now.

**Comment by Adda0** [comment]
We should merge the PR simply so that the current changes are finally on devel. We can add more PRs optimizing the current bottlenecks in the future. Once the experiments finish and there is no error, I would suggest merging.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 1]
```diff
@@ -850,15 +872,18 @@ namespace smt::noodler {
...
+                        len_formula = pr.second;
+                        found = true;
+                        break;
                     }
-                } else {
-                    this->axiomatized_instances.insert(refine, 1);
+                }
+                if(found) {
+                    block_curr_len(len_formula);
+                    return FC_CONTINUE;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:886]: Can you maybe explain what is happening here with some comments?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 2]
```diff
@@ -829,6 +829,28 @@ namespace smt::noodler {
         // Add dummy symbols for all disequations.
         // FIXME: we can possibly create more dummy symbols than the size of alphabet (196607 - from string theory standard), but it is edge-case that is nearly impossible to happen
         std::set<uint32_t> dummy_symbols{ util::get_dummy_symbols(std::max(new_symbs, size_t(3)), symbols_in_formula) };
+
+        // satisfiability checking via universality checking
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:833]: Can you explain this better? That we are doing universality checking as an optimization in the case that we have only one constraint that is x \not\in RL (if I understand correctly).

File: src/smt/theory_str_noodler/util.cc [hunk 3]
```diff
@@ -607,15 +624,33 @@ namespace smt::noodler::util {
             throw_error("unsupported operation in regex");
         }
 
+        STRACE("str-create_nfa",
+            tout << "--------------" << (make_complement ? "Complemented " : "") << "NFA for: " << mk_pp(const_cast<app*>(expression), const_cast<ast_manager&>(m)) << "---------------" << std::endl;
+        );
+
+        // intermediate automata reduction
+        if(nfa.size() < 1000) {
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/util.cc:632]: What is this magical constant?

**Comment by jurajsic** [comment]
Ok, I think we can merge

**Pull request merged**